README-VIS.md

```diff
@@ -0,0 +1,104 @@
+
+It is best for your `Dockerfile` to be under version control. You might include
+it in the repo for your project, and have Docker builds be
+triggered with each check in: This is what we've done in the Python and
+Shiny examples above. In other cases, because of build complexity, or because
+you're wrapping a third party tool, the real source code will be in a separate
+repo that you only reference: When that repo is updated, you'll need to bump
+a corresponding version number in your wrapper.
+
+## Inputs
+
+When your container starts up, an `INPUT_JSON_URL` environment variable will be available.
+(It is good practice and can make development easier to also check `INPUT_JSON`.)
+Its content will look something like [this](https://github.com/refinery-platform/docker_igv_js/blob/master/input_fixtures/good/input.json).
+- Input files are provided as a list of URLs under `file_relationships`.
+- Detailed metadata for each file in the dataset, not just your input, is provided
+in `node_info`.
+- User supplied strings or numbers are available in `parameters`.
+
+## Development
+
+To run your container locally, check out this project and add a description of your
+tool to `tools.py`, and start the demo with `python ./manage.py runserver`.
+The `image` field can reference images either on DockerHub, or in your local cache.
+
+## Gotchas
+
+There are a few things to be careful about when wrapping your tool.
+
+- **Port 80**: `django_docker_engine` assumes your tool will use port 80: If you're using
+a different port, that needs to be communicated when you publish your tool.
+(See the last section below.)
+- **Relocatable**: There will be a prefix in the path when your tool is served.
+No URLs you use should begin with `/`... but if you really need to know the 
+prefix, it is available in the `api_prefix` of your input.json.
+- **Keep the cookies**: Your tool will ultimately be run inside django, so the django
+session cookie needs to be preserved on any AJAX requests you make. Right now,
+FF and Chrome have slightly different implementations of the HTML5 Fetch API
+in this regard.
+- **No WebSockets**: WebSockets are a step beyond HTTP, and won't work across plain
+HTTP proxies like we use.
+- **No server state**: There are no provisions right
+now to preserve sessions or to make them available with restarts. Instead, try
+to update the query portion of the URL.
+- **Logging**: The Docker way to log is simply to output to STDOUT. If any files
+are updated or created, they should be listed in the `extra_directories` of your spec:
+If there is a lot of writing outside this space, you might exhaust the limited space
+dedicated to the container.
+- **Friendly error page**: Until your tool has fully started, it should either
+not reply, or reply with a non-200 response to requests for `/`. Once it has
+started, a 200 should be returned. If your tool can't start for some reason,
+it is still good to start a minimal HTTPD and return an error page with a 200
```

Comment:
error page with a 200?

Reply:
Yeah, but if this seems abhorrent could revisit. The idea is that django_docker only cares about one question: Is there a page I can show the end user? Maybe that page is your visualization... or maybe it's a message about why the visualization couldn't be started. In either case, all django_docker cares about is that the tool is in its final state and we can leave the please-wait page.

If we did it another way:
- The logic in django_docker would be more complicated.
- We'd need to decide on a status code sufficiently esoteric that we would not get it by mistake from a half-baked server... but not so weird that it would be hard for users to configure some HTTPD to serve this code.

... but I don't think I want to explain all of that here.

Reply:
rephrased:
> Until your tool has fully started, it should either
not reply, or reply with a non-200 response to requests for `/`. Once it has
started, a 200 should be returned. If your tool can't start for some reason,
please still respond with 200, but give an error message in the body instead.
Until requests for `/` return 200, `django_docker_engine` will keep a
please-wait page up.